File: Tests/iOS Tests/Tests/PNChannelGroupUnsubscribeTests.m
//
//  PNChannelGroupUnsubscribeTests.m
//  PubNub Tests
//
//  Created by Jordan Zucker on 6/28/15.
//
//

#import <PubNub/PubNub.h>

#import "PNBasicSubscribeTestCase.h"

static NSString * const kPNChannelGroupTestsName = @"PNChannelGroupUnsubscribeTests";

@interface PNChannelGroupUnsubscribeTests : PNBasicSubscribeTestCase

@end

@implementation PNChannelGroupUnsubscribeTests

- (BOOL)isRecording{
    return NO;
}

- (NSArray *)channelGroups {
    return @[
             kPNChannelGroupTestsName
             ];
}

- (void)setUp {
    [super setUp];
    [self performVerifiedRemoveAllChannelsFromGroup:kPNChannelGroupTestsName withAssertions:nil];
    PNWeakify(self);
    [self performVerifiedAddChannels:@[@"a", @"b"] toGroup:kPNChannelGroupTestsName withAssertions:^(PNAcknowledgmentStatus *status) {
        PNStrongify(self);
        XCTAssertNotNil(status);
        XCTAssertFalse(status.isError);
        XCTAssertEqual(status.operation, PNAddChannelsToGroupOperation);
        XCTAssertEqual(status.category, PNAcknowledgmentCategory);
        XCTAssertEqual(status.statusCode, 200);
    }];
    
    BOOL shouldObservePresence = NO;
    NSString *expectedMessage;
    NSNumber *expectedTimeToken;
    NSArray *expectedChannelGroups;
    if (self.invocation.selector == @selector(testSimpleUnsubscribeWithNoPresence)) {
        shouldObservePresence = NO;
        expectedMessage = @"********....... 8449 - 2015-12-22 13:28:00";
        expectedTimeToken = @14508196796692323;
        expectedChannelGroups = @[
                                  kPNChannelGroupTestsName
                                  ];
    } else if (self.invocation.selector == @selector(testSimpleUnsubscribeWithPresence)) {
        shouldObservePresence = YES;
        expectedMessage =  @"*********...... 8450 - 2015-12-22 13:28:01";
        expectedTimeToken = @14508196810395526;
        expectedChannelGroups = @[
                                  kPNChannelGroupTestsName,
                                  [kPNChannelGroupTestsName stringByAppendingString:@"-pnpres"]
                                  ];
    } else {
        XCTFail(@"unexpected presence state for channel group unsubscribe test");
    }
    
    self.didReceiveStatusAssertions = ^void (PubNub *client, PNSubscribeStatus *status) {
        PNStrongify(self);
        XCTAssertEqualObjects(self.client, client);
        XCTAssertNotNil(status);
        XCTAssertFalse(status.isError);
        XCTAssertEqual(status.category, PNConnectedCategory);
        XCTAssertEqual(status.subscribedChannels.count, 0);
        XCTAssertEqualObjects([NSSet setWithArray:status.subscribedChannelGroups],
                              [NSSet setWithArray:expectedChannelGroups]);
        XCTAssertEqual(status.operation, PNSubscribeOperation);
        NSLog(@"timeToken: %@", status.currentTimetoken);
        XCTAssertEqualObjects(status.currentTimetoken, expectedTimeToken);
        XCTAssertEqualObjects(status.currentTimetoken, status.data.timetoken);
        
    };
    self.didReceiveMessageAssertions = ^void (PubNub *client, PNMessageResult *message) {
        PNStrongify(self);
        XCTAssertEqualObjects(self.client, client);
        XCTAssertEqualObjects(client.uuid, message.uuid);
        XCTAssertNotNil(message.uuid);
        XCTAssertNil(message.authKey);
        XCTAssertEqual(message.statusCode, 200);
        XCTAssertTrue(message.TLSEnabled);
        XCTAssertEqual(message.operation, PNSubscribeOperation);
        NSLog(@"message:");
        NSLog(@"%@", message.data.message);
        XCTAssertEqualObjects(message.data.channel, @"a");
        XCTAssertEqualObjects(message.data.subscription, kPNChannelGroupTestsName);
        XCTAssertEqualObjects(message.data.message, expectedMessage);
        [self.channelGroupSubscribeExpectation fulfill];
        self.channelGroupSubscribeExpectation = nil;
    };
    
    [self PNTest_subscribeToChannelGroups:[self channelGroups] withPresence:shouldObservePresence];
}

- (void)testSimpleUnsubscribeWithPresence {
    PNWeakify(self);
    self.didReceiveStatusAssertions = ^void (PubNub *client, PNSubscribeStatus *status) {
        PNStrongify(self);
        XCTAssertNotNil(client);
        XCTAssertNotNil(status);
        XCTAssertEqualObjects(self.client, client);
        if (
            !(
              (status.category == PNDisconnectedCategory) ||
              (status.category == PNCancelledCategory)
              )
            ) {
            return;
        }
        XCTAssertEqual(status.category, PNDisconnectedCategory);
        XCTAssertFalse(status.isError);
        XCTAssertEqual(status.statusCode, 200);
        XCTAssertEqual(status.operation, PNUnsubscribeOperation);
        [self.channelGroupUnsubscribeExpectation fulfill];
    };
    [self PNTest_unsubscribeFromChannelGroups:[self channelGroups] withPresence:YES];
}

- (void)testSimpleUnsubscribeWithNoPresence {
    PNWeakify(self);
    self.didReceiveStatusAssertions = ^void (PubNub *client, PNSubscribeStatus *status) {
        PNStrongify(self);
        XCTAssertNotNil(client);
        XCTAssertNotNil(status);
        XCTAssertEqualObjects(self.client, client);
        if (
            !(
              (status.category == PNDisconnectedCategory) ||
              (status.category == PNCancelledCategory)
              )
            ) {
            return;
        }
        XCTAssertEqual(status.category, PNDisconnectedCategory);
        XCTAssertFalse(status.isError);
        XCTAssertEqual(status.statusCode, 200);
        XCTAssertEqual(status.operation, PNUnsubscribeOperation);
        [self.channelGroupUnsubscribeExpectation fulfill];
    };
    [self PNTest_unsubscribeFromChannelGroups:[self channelGroups] withPresence:NO];
}

@end

File: Tests/iOS Tests-Swift/PNChannelGroupUnsubscribeTests.swift
//
//  PNChannelGroupUnsubscribeTests.swift
//  PubNub Tests
//
//  Created by Vadim Osovets on 09/02/15.
//
//

import UIKit
import XCTest

class PNChannelGroupUnsubscribeTests: PNBasicSubscribeTestCase {
    
    let groupName = "PNChannelGroupUnsubscribeTests"
    
    override func isRecording() -> Bool {
        return false
    }
    
    lazy var channelGroups: [String] = {
        let groups = [self.groupName]
        return groups
        }()
    
    override func setUp() {
        super.setUp()
        
        self.performVerifiedRemoveAllChannelsFromGroup(self.groupName, assertions: nil)
        
        self.performVerifiedAddChannels(["a", "b"], channelGroup: self.groupName) { (status: PNAcknowledgmentStatus!) -> (Void) in
            XCTAssertNotNil(status)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.operation, PNOperationType.AddChannelsToGroupOperation)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.statusCode, 200)
        }
        
        // TODO: investigate me
        
//        self.didReceiveStatusAssertions = {(client: PubNub, status: PNSubscribeStatus) -> (Void) in
//            
//            XCTAssertEqualObjects(self.client, client)
//            XCTAssertNotNil(status)
//            XCTAssertFalse(status.error)
//            XCTAssertEqual(status.category, PNStatusCategory.PNConnectedCategory)
//            XCTAssertEqual(status.subscribedChannels.count, 0)
//            XCTAssertTrue(status.subscribedChannelGroups ==expectedChannelGroups, "Channel groups are not equal")
//            XCTAssertEqual(status.operation, PNSubscribeOperation)
//            
//            XCTAssertEqual(status.currentTimetoken, status.data.timetoken)
//        }
        
//        self.didReceiveMessageAssertions = {(client: PubNub!, message: PNMessageResult!) -> (Void) in
//            
//            XCTAssertEqualObjects(self.client, client);
//            XCTAssertEqualObjects(client.uuid, message.uuid);
//            XCTAssertNotNil(message.uuid);
//            XCTAssertNil(message.authKey);
//            XCTAssertEqual(message.statusCode, 200);
//            XCTAssertTrue(message.TLSEnabled);
//            XCTAssertEqual(message.operation, PNSubscribeOperation);
//            NSLog(@"message:");
//            NSLog(@"%@", message.data.message);
//            
//            self.channelGroupSubscribeExpectation.fulfill();
//        }
    }
    
    func testSimpleUnsubscribeWithPresence() {
        
        let shouldObservePresence = true
        let expectedMessage = "*****.......... 583 - 2015-06-28 21:52:31"
        let expectedTimeToken = 14355535508745522
        let expectedChannelGroups = ["PNChannelGroupUnsubscribeTests", "PNChannelGroupUnsubscribeTests-pnpres"]
        
        self.didReceiveStatusAssertions = {(client: PubNub, status: PNSubscribeStatus) -> (Void) in
            
            XCTAssertEqual(self.client, client)
            XCTAssertNotNil(status)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.category, PNStatusCategory.PNConnectedCategory)
            XCTAssertEqual(status.subscribedChannels.count, 0)
            
            // TODO: investigate
//            XCTAssertEqualObjects([NSSet setWithArray:status.subscribedChannelGroups],
//                [NSSet setWithArray:expectedChannelGroups]);
//            XCTAssertEqual(status.operation, PNSubscribeOperation);
            
            //        XCTAssertEqualObjects(status.currentTimetoken, expectedTimeToken);
            XCTAssertEqual(status.currentTimetoken, status.data.timetoken)
            
        };
        self.didReceiveMessageAssertions = {(client: PubNub!, message: PNMessageResult!) -> (Void) in
            
            XCTAssertEqual(self.client, client)
            XCTAssertEqual(client.uuid(), message.uuid)
            XCTAssertNotNil(message.uuid)
            XCTAssertNil(message.authKey)
            XCTAssertEqual(message.statusCode, 200)
            XCTAssertTrue(message.TLSEnabled)
            XCTAssertEqual(message.operation, PNOperationType.SubscribeOperation)
            
            //        XCTAssertEqualObjects(message.data.message, expectedMessage);
            
            self.channelGroupSubscribeExpectation.fulfill()
        };
        
        self.PNTest_subscribeToChannelGroups(self.channelGroups, presense:shouldObservePresence)
        
        self.didReceiveStatusAssertions = {(client: PubNub, status: PNSubscribeStatus) -> (Void) in
            
            XCTAssertNotNil(client)
            XCTAssertNotNil(status)
            XCTAssertEqual(self.client, client)
//            XCTAssertEqual(status.category, PNStatusCategory.PNDisconnectedCategory)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.statusCode, 200)
//            XCTAssertEqual(status.operation, PNOperationType.UnsubscribeOperation)
            
            self.channelGroupUnsubscribeExpectation.fulfill();
        }

        self.PNTest_unsubscribeFromChannelGroups(self.channelGroups, presense: true)
    }
    
    func testSimpleUnsubscribeWithNoPresence() {
        
        let shouldObservePresence = false
        let expectedMessage = "*****.......... 583 - 2015-06-28 21:52:31"
        let expectedTimeToken = 14355535508745522
        let expectedChannelGroups = ["PNChannelGroupUnsubscribeTests"]
        
        
        self.didReceiveMessageAssertions = {(client: PubNub!, message: PNMessageResult!) -> (Void) in
            
            XCTAssertEqual(self.client, client)
            XCTAssertEqual(client.uuid(), message.uuid)
            XCTAssertNotNil(message.uuid)
            XCTAssertNil(message.authKey)
            XCTAssertEqual(message.statusCode, 200)
            XCTAssertTrue(message.TLSEnabled)
            XCTAssertEqual(message.operation, PNOperationType.SubscribeOperation)
            
            //        XCTAssertEqualObjects(message.data.message, expectedMessage);
            
            self.channelGroupSubscribeExpectation.fulfill()
        };
        
        self.PNTest_subscribeToChannelGroups(self.channelGroups, presense:shouldObservePresence)
        
        self.didReceiveStatusAssertions = {(client: PubNub, status: PNSubscribeStatus) -> (Void) in
            
            XCTAssertNotNil(client)
            XCTAssertNotNil(status)
            XCTAssertEqual(self.client, client)
            
            // TODO: try to investigate it
            
//            XCTAssertEqual(status.category, PNStatusCategory.PNDisconnectedCategory)
            XCTAssertFalse(status.error)
            
            XCTAssertEqual(status.statusCode, 200)
//            XCTAssertEqual(status.operation, PNOperationType.UnsubscribeOperation)
            
            self.channelGroupUnsubscribeExpectation.fulfill();
        }
        
        self.PNTest_unsubscribeFromChannelGroups(self.channelGroups, presense: false)
    }
}
